src/vcon/MarketGovernance.sol

```diff
@@ -0,0 +1,653 @@
+        address destination = underlyingTokenToHoldingDeposit[denomination];
+        require(
+            destination != address(0),
+            "MarketGovernance: invalid destination"
+        );
+
+        /// ---------- Effects ----------
+
+        _accrue(source); /// update profitPerVCON in the source so the user gets paid out at the current share price
+
+        /// figure out how balanced the system is before withdraw
+
+        /// amount of PCV to withdraw is the amount vcon * venue balance / total vcon staked on venue
+        uint256 amountPcv = getProRataPCVAmounts(source, shareAmount);
+
+        /// user updates
+        venueUserShares[source][msg.sender] -= shareAmount;
+
+        /// venue updates
+        venueTotalShares[source] -= shareAmount;
+
+        /// ---------- Interactions ----------
+
+        {
+            PCVRouter(pcvRouter).movePCV(
+                source,
+                destination,
+                address(0),
+                amountPcv,
+                denomination,
+                denomination
+            );
+        }
+
+        {
+            uint256 amountVcon = sharesToVcon(source, shareAmount);
+            vcon().safeTransfer(vconRecipient, amountVcon); /// transfer VCON amount to recipient
+            emit Unstaked(source, msg.sender, amountVcon, amountPcv);
+        }
+    }
+
+    /// @notice rebalance PCV without staking or unstaking VCON
+    /// each individual action must make the system more balanced
+    /// as a whole, otherwise it will revert
+    /// @param movements information on all pcv movements
+    /// including sources, destinations, amounts and swappers
+    function rebalance(
+        Rebalance[] calldata movements
+    ) external globalLock(1) whenNotPaused {
+        /// read unsafe because we are at lock level 1
+        IPCVOracle oracle = pcvOracle();
+        uint256 totalPcv = oracle.getTotalPcvUnsafe();
+        uint256 totalVconStaked = getTotalVconStaked();
+        unchecked {
+            for (uint256 i = 0; i < movements.length; i++) {
+                address source = movements[i].source;
+                address destination = movements[i].destination;
+                address swapper = movements[i].swapper;
+                uint256 amountPcv = movements[i].amountPcv;
+
+                /// validate source, dest and swapper
+                require(
+                    oracle.isVenue(source),
+                    "MarketGovernance: invalid source"
+                );
+                require(
+                    oracle.isVenue(destination),
+                    "MarketGovernance: invalid destination"
+                );
+                /// if swapper is used, validate in PCV Router whiteliste
+                if (swapper != address(0)) {
+                    require(
+                        PCVRouter(pcvRouter).isPCVSwapper(swapper),
+                        "MarketGovernance: invalid swapper"
+                    );
+                }
```

Comment:
This is all done by the PCVRouter already (and the PCVRouter also checks that the swapper can swap between the underlying tokens), you can remove the checks

Reply:
The system entry point is where we should be sanitizing inputs if we want to follow best practices. It's not a clean pattern to have an internal contract of the system do the sanitization after passing through an external contract.

Reply:
I think it's fine to rely on the behavior of our other protocol contracts, but you could add a public view function to the PCVRouter that do all the checks (and use it in both places), if you really want to check before. Or use the unchecked function if you're sure all inputs are checked already. But implementing the same logic in multiple places is not good practice either (you're also not doing all the checks, the swapper should be called to know if it can swap between the underlying tokens)